art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -0,0 +1,428 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements Sleeper Agent attack on Neural Networks.
+
+| Paper link: https://arxiv.org/abs/2106.08970
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Any, Tuple, TYPE_CHECKING, List
+import random
+
+import numpy as np
+from tqdm.auto import trange
+
+
+from art.attacks.poisoning import GradientMatchingAttack
+
+if TYPE_CHECKING:
+    # pylint: disable=C0412
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+    import torch
+
+logger = logging.getLogger(__name__)
+
+
+class SleeperAgentAttack(GradientMatchingAttack):
+    """
+    Implementation of Sleeper Agent Attack
+
+    | Paper link: https://arxiv.org/pdf/2106.08970.pdf
+    """
+
+    def __init__(
+        self,
+        classifier: "CLASSIFIER_NEURALNETWORK_TYPE",
+        percent_poison: float,
+        patch: np.ndarray,
+        epsilon: float = 0.1,
+        max_trials: int = 8,
+        max_epochs: int = 250,
+        learning_rate_schedule: Tuple[List[float], List[int]] = ([1e-1, 1e-2, 1e-3, 1e-4], [100, 150, 200, 220]),
+        batch_size: int = 128,
+        clip_values: Tuple[float, float] = (0, 1.0),
+        verbose: int = 1,
+        indices_target: List[int] = None,
+        patching_strategy: str = "random",
+        selection_strategy: str = "random",
+        retraining_factor: int = 1,
+        model_retrain: bool = False,
+        model_retraining_epoch: int = 1,
+        class_source: int = 0,
+        class_target: int = 1,
+    ):
+        """
+        Initialize a Sleeper Agent poisoning attack.
+
+        :param classifier: The proxy classifier used for the attack.
```

Comment:
Pleas order the param descriptions according to arguments in method.

Reply:
Done